src/shell/braces.zig

```diff
@@ -0,0 +1,805 @@
+            .open => {
+                try nested_brace_stack.push(0);
+            },
+            .comma => {
+                const val = nested_brace_stack.topPtr().?;
+                val.* += 1;
+                prev_comma = true;
+            },
+            .close => {
+                var variants = nested_brace_stack.pop().?;
+                if (!prev_comma) {
+                    variants += 1;
+                }
+                if (nested_brace_stack.len > 0) {
+                    const top = nested_brace_stack.topPtr().?;
+                    top.* += variants - 1;
+                } else if (variant_count == 0) {
+                    variant_count = variants;
+                } else {
+                    variant_count *= variants;
+                }
+            },
+            else => {},
+        }
+    }
+
+    return variant_count;
+}
+
+pub fn buildExpansionTableAlloc(alloc: Allocator, tokens: []Token) !std.ArrayList(ExpansionVariant) {
+    var table = std.ArrayList(ExpansionVariant).init(alloc);
+    try buildExpansionTable(tokens, &table);
+    return table;
+}
+
+pub fn buildExpansionTable(
+    tokens: []Token,
+    table: *std.ArrayList(ExpansionVariant),
+) !void {
+    const BraceState = struct {
+        tok_idx: u16,
+        variants: u16,
+        prev_tok_end: u16,
+    };
+    var brace_stack = StackStack(BraceState, u4, MAX_NESTED_BRACES){};
+
+    var i: u16 = 0;
+    var prev_close = false;
+    while (i < tokens.len) : (i += 1) {
+        switch (tokens[i]) {
+            .open => {
+                const table_idx: u16 = @intCast(table.items.len);
+                tokens[i].open.idx = table_idx;
+                try brace_stack.push(.{
+                    .tok_idx = i,
+                    .variants = 0,
+                    .prev_tok_end = i,
+                });
+            },
+            .close => {
+                var top = brace_stack.pop().?;
+
+                try table.append(.{
+                    .end = i,
+                    .start = top.prev_tok_end + 1,
+                });
+
+                top.prev_tok_end = i;
+                top.variants += 1;
+
+                tokens[top.tok_idx].open.end = @intCast(table.items.len);
+                prev_close = true;
+            },
+            .comma => {
+                var top = brace_stack.topPtr().?;
+
+                try table.append(.{
+                    .end = i,
+                    .start = top.prev_tok_end + 1,
+                });
+
+                prev_close = false;
+
+                top.prev_tok_end = i;
+                top.variants += 1;
+            },
+            else => {
+                prev_close = false;
+            },
+        }
+    }
+
+    if (bun.Environment.allow_assert) {
+        for (table.items[0..], 0..) |variant, kdjsd| {
+            _ = kdjsd;
+            std.debug.assert(variant.start != 0 and variant.end != 0);
+        }
+    }
+}
+
+const Encoding = @import("./shell.zig").StringEncoding;
```

Comment:
It would be reasonable to add `bun.shell` and use that instead of re-declaring the import a bunch of times in this file

Reply:
There are footguns with `@import`: case insensitive filesystems don't work consistently, if we move files around it's tedious to rewrite all the references since that tooling in Zig doesn't exist yet, etc